src/vm/eventpipebuffermanager.cpp

```diff
@@ -0,0 +1,595 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "eventpipeconfiguration.h"
+#include "eventpipebuffer.h"
+#include "eventpipebuffermanager.h"
+
+#ifdef FEATURE_PERFTRACING
+
+EventPipeBufferManager::EventPipeBufferManager()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    m_pPerThreadBufferList = new SList<SListElem<EventPipeBufferList*>>();
+    m_sizeOfAllBuffers = 0;
+    m_lock.Init(LOCK_TYPE_DEFAULT);
+
+#ifdef _DEBUG
+    m_numBuffersAllocated = 0;
+    m_numBuffersStolen = 0;
+#endif // _DEBUG
+}
+
+EventPipeBuffer* EventPipeBufferManager::AllocateBufferForThread(Thread *pThread, unsigned int requestSize)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_NOTRIGGER;
+        MODE_ANY;
+        PRECONDITION(pThread != NULL);
+        PRECONDITION(requestSize > 0);
+    }
+    CONTRACTL_END;
+
+    // Allocating a buffer requires us to take the lock.
+    SpinLockHolder _slh(&m_lock);
```

Comment:
It turns out that I will need more to happen under the spin lock - this PR doesn't include the clean-up routines, and so they will require list from the thread object to be under lock.  We could potentially try not to lock on the allocations, but we'd have to allocate pre-emptively and then free it if we didn't need it.

I'm going to be closing this PR out because I have more changes on top of these and it makes sense to review them together.  I will see if there is anything that I can do here.